File: test/goldens/charts/sleep_schedule_chart_golden_test.dart
@Tags(['golden'])
library;

import 'package:flutter_test/flutter_test.dart';

import 'package:openvitals/core/period/time_range.dart';
import 'package:openvitals/core/time/local_date.dart';
import 'package:openvitals/domain/model/sleep_models.dart';
import 'package:openvitals/features/sleep/application/sleep_display.dart';
import 'package:openvitals/features/sleep/presentation/sleep_schedule_chart.dart';

import '../../support/golden_harness.dart';

/// [SleepScheduleStageChart] — a week of nights on one clock.
///
/// Its whole reason for existing is the 18:00 anchor: a night that starts at
/// 23:40 and ends at 07:10 crosses midnight, and on a plain 00:00-24:00 axis it
/// would be drawn as two bars at opposite ends of the chart. Anchored at six in
/// the evening, it is one bar. Nothing but a picture shows whether that still
/// holds, and the fixture leans on it — every night here straddles midnight, and
/// one of them (Saturday) goes to bed AFTER it, which is the case that wraps.
///
/// This is also the one chart in the app whose label gutter is on the RIGHT, so
/// its x-axis row is padded on the right rather than inset on the left. The
/// golden is the only thing that would catch someone "fixing" that.
void main() {
  // The week ending on the golden day. `date` is the night's DATE — the morning
  // you woke up — which is why every in-bed window starts on the day before.
  const monday = LocalDate(2026, 6, 16);

  SleepStage stage(int type, DateTime start, Duration length) => SleepStage(
        startTime: start,
        endTime: start.add(length),
        stageType: type,
      );

  /// A night with plausible architecture, laid down from its bedtime.
  List<SleepStage> stagesFrom(DateTime start, Duration total) {
    final stages = <SleepStage>[];
    var cursor = start;
    const cycle = <(int, int)>[
      (SleepStage.stageAwake, 8),
      (SleepStage.stageLight, 45),
      (SleepStage.stageDeep, 50),
      (SleepStage.stageLight, 35),
      (SleepStage.stageRem, 40),
    ];
    var index = 0;
    while (cursor.isBefore(start.add(total))) {
      final (type, minutes) = cycle[index % cycle.length];
      final length = Duration(minutes: minutes);
      final end = cursor.add(length);
      stages.add(
        stage(type, cursor, end.isAfter(start.add(total))
            ? start.add(total).difference(cursor)
            : length),
      );
      cursor = end;
      index++;
    }
    return stages;
  }

  SleepScheduleDay night(
    LocalDate date,
    int bedHour,
    int bedMinute,
    Duration length,
  ) {
    // Bedtime belongs to the EVENING BEFORE the night's date, unless it is past
    // midnight already — the Saturday lie-in below.
    final start = bedHour >= 12
        ? DateTime(date.year, date.month, date.day, bedHour, bedMinute)
            .subtract(const Duration(days: 1))
        : DateTime(date.year, date.month, date.day, bedHour, bedMinute);
    return SleepScheduleDay(
      date: date,
      inBedStart: start,
      inBedEnd: start.add(length),
      stages: stagesFrom(start, length),
    );
  }

  final week = <SleepScheduleDay>[
    night(monday, 23, 20, const Duration(hours: 7, minutes: 40)),
    night(monday.plusDays(1), 23, 5, const Duration(hours: 7, minutes: 55)),
    night(monday.plusDays(2), 0, 10, const Duration(hours: 6, minutes: 50)),
    // Thursday: the tracker recorded a window but no stages at all. The chart
    // draws a solid bar rather than an empty slot — "I know when, not what".
    SleepScheduleDay(
      date: monday.plusDays(3),
      inBedStart: DateTime(2026, 6, 18, 23, 45),
      inBedEnd: DateTime(2026, 6, 19, 7, 0),
    ),
    night(monday.plusDays(4), 23, 55, const Duration(hours: 6, minutes: 30)),
    // Saturday: a late night out, in bed at 01:05, up at 09:40. Past the 18:00
    // anchor by more than a day's worth of minutes — this is the wrap case.
    night(monday.plusDays(5), 1, 5, const Duration(hours: 8, minutes: 35)),
    // Sunday: no session at all. A missing night is a gap, not a zero.
    const SleepScheduleDay(
      date: LocalDate(2026, 6, 22),
      inBedStart: null,
      inBedEnd: null,
    ),
  ];

  testWidgets('a week of nights', (tester) async {
    await expectChartGoldenBothThemes(
      tester,
      () => SleepScheduleStageChart(
        title: 'Sleep schedule',
        summaryText: 'This week · 7h 24m avg',
        days: week,
        selectedRange: TimeRange.week,
      ),
      name: 'sleep_schedule_chart_week',
    );
  });

  testWidgets('a week with the average bedtime and wake-up marked',
      (tester) async {
    await expectChartGoldenBothThemes(
      tester,
      () => SleepScheduleStageChart(
        title: 'Sleep schedule',
        summaryText: 'This week · 7h 24m avg',
        days: week,
        selectedRange: TimeRange.week,
        // Minutes of the day, not anchored minutes: 23:35 to bed, up at 07:20.
        // The chart anchors them itself, which is the only reason a bedtime of
        // 1415 and a wake-up of 440 can share one axis.
        averageSchedule: const SleepOverviewSchedule(1415, 440),
        selectedDate: monday.plusDays(4),
        onDateSelected: (_) {},
      ),
      name: 'sleep_schedule_chart_markers',
    );
  });
}

File: app/src/androidTest/kotlin/tech/mmarca/openvitals/features/sleep/SleepScheduleChartGoldenTest.kt
package tech.mmarca.openvitals.features.sleep

import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.ui.Modifier
import androidx.compose.ui.test.junit4.v2.createComposeRule
import androidx.compose.ui.unit.dp
import java.time.Instant
import java.time.LocalDate
import org.junit.Rule
import org.junit.Test
import tech.mmarca.openvitals.core.period.DatePeriod
import tech.mmarca.openvitals.core.period.TimeRange
import tech.mmarca.openvitals.core.presentation.DateTimeFormatterProvider
import tech.mmarca.openvitals.domain.model.SleepStage
import tech.mmarca.openvitals.domain.preferences.SleepWindow
import tech.mmarca.openvitals.testing.OpenVitalsVisualTestSurface
import tech.mmarca.openvitals.testing.assertVisualRootMatchesGolden
import tech.mmarca.openvitals.testing.goldenInstant

/**
 * [SleepScheduleStageChart]: a week of nights on one clock anchored at 18:00, so a night
 * crossing midnight is one bar. Every night here straddles midnight and Saturday goes to bed after it.
 * This is the one chart with its label gutter on the right.
 */
class SleepScheduleChartGoldenTest {

    @get:Rule
    val composeRule = createComposeRule()

    @Test
    fun aWeekOfNights() {
        composeRule.setContent {
            OpenVitalsVisualTestSurface(width = 360.dp, height = 400.dp) {
                ScheduleChart(days = week(), summary = "This week · 7h 24m avg")
            }
        }

        composeRule.assertVisualRootMatchesGolden("sleep_schedule_chart_week")
    }

    @Test
    fun aPartlyStagedNightReadsAtItsFullDuration() {
        // A full night the tracker staged only near the end. The bar draws the whole time in bed
        // as a base block with the staged tail overlaid. The middle night is fully staged.
        composeRule.setContent {
            OpenVitalsVisualTestSurface(width = 360.dp, height = 400.dp) {
                ScheduleChart(
                    days = listOf(
                        tailStaged(MONDAY),
                        night(MONDAY.plusDays(1), bedHour = 23, bedMinute = 5, minutes = 475),
                        tailStaged(MONDAY.plusDays(2)),
                    ),
                    summary = "This week · partly staged",
                    period = DatePeriod(MONDAY, MONDAY.plusDays(2)),
                )
            }
        }

        composeRule.assertVisualRootMatchesGolden("sleep_schedule_chart_partly_staged")
    }

    @Test
    fun aWeekWithTheAverageBedtimeAndWakeUpMarked() {
        // Minutes of the day, not anchored minutes. The chart anchors them itself.
        composeRule.setContent {
            OpenVitalsVisualTestSurface(width = 360.dp, height = 400.dp) {
                ScheduleChart(
                    days = week(),
                    summary = "This week · 7h 24m avg",
                    averageSchedule = SleepOverviewSchedule(startMinute = 1_415, endMinute = 440),
                    selectedDate = MONDAY.plusDays(4),
                    onDateSelected = {},
                )
            }
        }

        composeRule.assertVisualRootMatchesGolden("sleep_schedule_chart_markers")
    }

    @androidx.compose.runtime.Composable
    private fun ScheduleChart(
        days: List<SleepScheduleDay>,
        summary: String,
        period: DatePeriod = WEEK,
        averageSchedule: SleepOverviewSchedule? = null,
        selectedDate: LocalDate? = null,
        onDateSelected: ((LocalDate) -> Unit)? = null,
    ) {
        SleepScheduleStageChart(
            title = "Sleep schedule",
            summaryText = summary,
            days = days,
            sleepWindow = SleepWindow.Default,
            selectedRange = TimeRange.WEEK,
            period = period,
            dateTimeFormatterProvider = DateTimeFormatterProvider(),
            averageSchedule = averageSchedule,
            selectedDate = selectedDate,
            onDateSelected = onDateSelected,
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp),
        )
    }

    private companion object {
        // `date` is the morning you woke up, so every in-bed window starts the day before.
        val MONDAY: LocalDate = LocalDate.of(2026, 6, 16)
        val WEEK = DatePeriod(MONDAY, LocalDate.of(2026, 6, 22))

        /** A night with plausible architecture, laid down from its bedtime. */
        fun stagesFrom(start: Instant, totalMinutes: Long): List<SleepStage> {
            val cycle = listOf(
                SleepStage.STAGE_AWAKE to 8L,
                SleepStage.STAGE_LIGHT to 45L,
                SleepStage.STAGE_DEEP to 50L,
                SleepStage.STAGE_LIGHT to 35L,
                SleepStage.STAGE_REM to 40L,
            )
            val end = start.plusSeconds(totalMinutes * 60L)
            val stages = mutableListOf<SleepStage>()
            var cursor = start
            var index = 0
            while (cursor.isBefore(end)) {
                val (type, minutes) = cycle[index % cycle.size]
                val stageEnd = minOf(cursor.plusSeconds(minutes * 60L), end)
                stages += SleepStage(startTime = cursor, endTime = stageEnd, stageType = type)
                cursor = stageEnd
                index++
            }
            return stages
        }

        fun night(
            date: LocalDate,
            bedHour: Int,
            bedMinute: Int,
            minutes: Long,
        ): SleepScheduleDay {
            // Bedtime belongs to the evening before, unless it is past midnight already.
            val bedDate = if (bedHour >= 12) date.minusDays(1) else date
            val start = goldenInstant(
                bedDate.year,
                bedDate.monthValue,
                bedDate.dayOfMonth,
                bedHour,
                bedMinute,
            )
            return SleepScheduleDay(
                date = date,
                inBedStart = start,
                inBedEnd = start.plusSeconds(minutes * 60L),
                stages = stagesFrom(start, minutes),
            )
        }

        fun tailStaged(date: LocalDate): SleepScheduleDay {
            val bed = goldenInstant(
                date.minusDays(1).year,
                date.minusDays(1).monthValue,
                date.minusDays(1).dayOfMonth,
                23,
                30,
            )
            val wake = bed.plusSeconds(450L * 60L)
            return SleepScheduleDay(
                date = date,
                inBedStart = bed,
                inBedEnd = wake,
                // ~50 min of stages at the tail of a 7h30m night — well under half.
                stages = stagesFrom(wake.minusSeconds(50L * 60L), 50L),
            )
        }

        fun week(): List<SleepScheduleDay> = listOf(
            night(MONDAY, bedHour = 23, bedMinute = 20, minutes = 460),
            night(MONDAY.plusDays(1), bedHour = 23, bedMinute = 5, minutes = 475),
            night(MONDAY.plusDays(2), bedHour = 0, bedMinute = 10, minutes = 410),
            // Thursday: a window but no stages. The chart draws a solid bar.
            SleepScheduleDay(
                date = MONDAY.plusDays(3),
                inBedStart = goldenInstant(2026, 6, 18, 23, 45),
                inBedEnd = goldenInstant(2026, 6, 19, 7, 0),
            ),
            night(MONDAY.plusDays(4), bedHour = 23, bedMinute = 55, minutes = 390),
            // Saturday: in bed at 01:05, up at 09:40. The wrap case.
            night(MONDAY.plusDays(5), bedHour = 1, bedMinute = 5, minutes = 515),
            // Sunday: no session at all. A missing night is a gap, not a zero.
            SleepScheduleDay(
                date = LocalDate.of(2026, 6, 22),
                inBedStart = null,
                inBedEnd = null,
            ),
        )
    }
}
